src/account/UpgradeableModularAccount.sol

```diff
@@ -340,83 +344,72 @@ contract UpgradeableModularAccount is
         UserOperation calldata userOp,
         bytes32 userOpHash
     ) internal returns (uint256 validationData) {
-        if (userOpValidationFunction.isEmpty()) {
+        if (userOpValidationFunction.isEmptyOrMagicValue()) {
+            // If the validation function is empty, then the call cannot proceed.
+            // Alternatively, the validation function may be set to the RUNTIME_VALIDATION_ALWAYS_ALLOW magic
+            // value, in which case we also revert.
             revert UserOpValidationFunctionMissing(selector);
         }
 
         uint256 currentValidationData;
 
         // Do preUserOpValidation hooks
-        EnumerableMap.Bytes32ToUintMap storage preUserOpValidationHooks =
+        EnumerableSet.Bytes32Set storage preUserOpValidationHooks =
             getAccountStorage().selectorData[selector].preValidationHooks;
 
         uint256 preUserOpValidationHooksLength = preUserOpValidationHooks.length();
         for (uint256 i = 0; i < preUserOpValidationHooksLength; ++i) {
-            (bytes32 key,) = preUserOpValidationHooks.at(i);
+            bytes32 key = preUserOpValidationHooks.at(i);
             FunctionReference preUserOpValidationHook = _toFunctionReference(key);
 
-            if (!preUserOpValidationHook.isEmptyOrMagicValue()) {
-                (address plugin, uint8 functionId) = preUserOpValidationHook.unpack();
-                currentValidationData = IPlugin(plugin).preUserOpValidationHook(functionId, userOp, userOpHash);
+            (address plugin, uint8 functionId) = preUserOpValidationHook.unpack();
+            currentValidationData = IPlugin(plugin).preUserOpValidationHook(functionId, userOp, userOpHash);
 
-                if (uint160(currentValidationData) > 1) {
-                    // If the aggregator is not 0 or 1, it is an unexpected value
-                    revert UnexpectedAggregator(plugin, functionId, address(uint160(currentValidationData)));
-                }
-                validationData = _coalescePreValidation(validationData, currentValidationData);
-            } else {
-                // Function reference cannot be 0 and _RUNTIME_VALIDATION_ALWAYS_ALLOW is not permitted here.
-                revert InvalidConfiguration();
+            if (uint160(currentValidationData) > 1) {
+                // If the aggregator is not 0 or 1, it is an unexpected value
+                revert UnexpectedAggregator(plugin, functionId, address(uint160(currentValidationData)));
             }
+            validationData = _coalescePreValidation(validationData, currentValidationData);
         }
 
         // Run the user op validationFunction
         {
```

Comment:
Just wondering if this block scope is still necessary

Reply:
The block scope is done not for stack depth reasons, but to prevent the variables `(address plugin, uint8 functionId)` declared in the loop executing pre user op validation hooks from shadowing the same variables used when unpacking the user op validation function. Do you think it's worth renaming them and removing the block scope?

Reply:
Ah okay, that seems fine. Just generally prefer less nesting when possible